### PR TITLE
Suggestion Cog

### PR DESCRIPTION
**Pull request by Duued**
This update will add a suggestion cog

Users can use `[p] suggest content`, then they will be sent a View with a select to choose what type of suggestion this is.

This also updated config.toml and config.template.toml to allow for it to be set by anyone that hosts the bot.

### CODE REVIEW

File: modules/info.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    :param entity: The Member, User, Role, TextChannel, or Guild to get information about.
+    """
+
+    def __init__(
+        self,
+        *,
+        author: Union[discord.Member, discord.User],
+        entity: Union[discord.Member, discord.User, discord.Role, discord.TextChannel, discord.Guild]
+        # author: Union[discord.Member, discord.User],
+        # entity: Union[discord.Member, discord.User, discord.Role, discord.TextChannel, discord.Guild],
```
Comment (AbstractUmbra) [inline modules/info.py:48]: Commented parameters?

File: config.template.toml [hunk 2]
```diff
@@ -24,6 +24,9 @@ webhook_avatar_url = "" # optional
 runner = 123456789 # optional: sets the webhook avatar url (will be overridden by the webhook_avatar_url attribute)
 level = 20
 
+[SUGGESTIONS]
```
Comment (AbstractUmbra) [inline config.template.toml:27]: Mark the parent key optional too, if it is.

File: modules/info.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    ):
+        super().__init__()
+        created_at: str = f"{discord.utils.format_dt(entity.created_at)} ({discord.utils.format_dt(entity.created_at, 'R')})"
+        if isinstance(entity, discord.Member) and entity.joined_at:
+            joined_at = f"\n\nJoined At: {discord.utils.format_dt(entity.joined_at)} ({discord.utils.format_dt(entity.joined_at, 'R')})"
+        else:
+            joined_at = ""
+
+        start: str = f"Name: {entity.name}\n\nID: {entity.id}\n\nCreated At: {created_at}{joined_at}"
+        description = start
```
Comment (AbstractUmbra) [inline modules/info.py:58]: Why are we doing `description = start` if `start` exists? Just rename it...?

File: modules/info.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        Args:
+            entity: The user, role, or TextChannel to get information about"""
+        embed = InformationEmbed(author=ctx.author, entity=entity or ctx.author)
+        await ctx.send(embed=embed)
+
+    @info.command(name="guild", brief="Get the current guild's information.")
+    @commands.guild_only()
+    async def guild_info(self, ctx: core.Context):
+        if ctx.guild is None:
+            raise commands.CheckFailure("This command must be ran in a guild.")  # Make pyright stop throwing fits
```
Comment (AbstractUmbra) [inline modules/info.py:103]: Annotate `ctx` as `core.GuildContext` since we are guarded correctly with the `guild_only()` decorated check. This removes the need for lines 102-103.

```suggestion
    @commands.guild_only()
    async def guild_info(self, ctx: core.GuildContext) -> None:
```

File: modules/suggestions.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+        if self.message:
+            await self.message.delete()
+
+
+class TypeView(ui.View):
+    def __init__(self, *, original_author: Union[discord.Member, discord.User], suggestion: str, webhook: discord.Webhook):
+        super().__init__(timeout=180)
+        self.original_author = original_author
+        self.add_item(TypeSelect(original_author=original_author, suggestion=suggestion, webhook=webhook))
+        self.message: discord.Message
```
Comment (AbstractUmbra) [inline modules/suggestions.py:100]: The more accurate thing to do here is to move `message` as a class level type annotation and assign there. That way we don't need it as a parameter since we technically duck-type it in.

```suggestion
class TypeView(ui.View):
    message: discord.Message | discord.WebhookMessage
```

File: modules/suggestions.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    @commands.group(name="suggest", brief="Send a suggestion for the server, or a library.", invoke_without_command=True)
+    async def suggest(self, ctx: core.Context, *, suggestion: str | None = None):
+        if not suggestion:
+            return await ctx.reply("Re-execute the command. Make sure to give your suggestion this time!")
+        view = TypeView(original_author=ctx.author, suggestion=suggestion, webhook=self.webhook)
+        message = await ctx.send(
+            "Please select the type of suggestion this is. Your suggestion will not be sent until this step is completed.",
+            view=view,
+        )
+        view.message = message
```
Comment (AbstractUmbra) [inline modules/suggestions.py:121]:
```suggestion
        view.message = await ctx.send(
            "Please select the type of suggestion this is. Your suggestion will not be sent until this step is completed.",
            view=view,
        )
```

**Review by AbstractUmbra — CHANGES_REQUESTED** [review]
Hey! Thanks for the PR. Some minor changes here however...

File: modules/info.py (hunk 1 above)
Comment (Duued) [reply modules/info.py:48]: Those were old and once again something else I forgot to remove

File: modules/suggestions.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+            )
+        author = self.original_author
+        if self.values[0] == "1":  # Guild Suggestion
+            suggestion_type = "the guild"
+        elif self.values[0] == "2":  # Pythonistabot Suggestion
+            suggestion_type = "Pythonistabot"
+        elif self.values[0] == "3":
+            suggestion_type = "TwitchIO"
+        else:
+            suggestion_type = "Wavelink"
```
Comment (AbstractUmbra) [inline modules/suggestions.py:83]: This could be a dict lookup with a default

File: modules/suggestions.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        embed.set_footer(text=f"Suggestion sent by {author} (ID: {author.id})")
+        await self.webhook.send(embed=embed, avatar_url=author.display_avatar, username=author.name)
+        if self.message:
+            await self.message.delete()
+
+
+class TypeView(ui.View):
+    def __init__(self, *, original_author: Union[discord.Member, discord.User], suggestion: str, webhook: discord.Webhook):
+        super().__init__(timeout=180)
+        self.message: discord.Message | discord.WebhookMessage
```
Comment (AbstractUmbra) [inline modules/suggestions.py:96]:
```suggestion
class TypeView(ui.View):
    self.message: discord.Message | discord.WebhookMessage

    def __init__(self, *, original_author: Union[discord.Member, discord.User], suggestion: str, webhook: discord.Webhook):
        super().__init__(timeout=180)
```

**Review by AbstractUmbra — CHANGES_REQUESTED** [review]
One last tidbit

File: modules/suggestions.py (hunk 8 above)
Comment (Duued) [reply modules/suggestions.py:96]: Okay I'll do it, but what difference will it make?

File: launcher.py
Comment (AbstractUmbra) [inline launcher.py:1]: Why was this file deleted?

File: config.template.toml [hunk 9]
```diff
@@ -24,6 +24,9 @@ webhook_avatar_url = "" # optional
 runner = 123456789 # optional: sets the webhook avatar url (will be overridden by the webhook_avatar_url attribute)
 level = 20
 
+[SUGGESTIONS] # optional
+url = "" # Webhook URL - optional
```
Comment (AbstractUmbra) [inline config.template.toml:28]: The key isn't optional, the parent key is.
Also make the config keys sort-of sane by naming;-
```suggestion
webhook_url = ""
```

File: modules/suggestions.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        view = TypeView(original_author=ctx.author, suggestion=suggestion, webhook=self.webhook)
+        view.message = await ctx.send(
+            "Please select the type of suggestion this is. Your suggestion will not be sent until this step is completed.",
+            view=view,
+        )
+
+
+async def setup(bot: core.Bot):
+    if key := core.CONFIG.get("SUGGESTIONS"):
+        await bot.add_cog(Suggestions(bot, key["url"]))
```
Comment (AbstractUmbra) [inline modules/suggestions.py:126]:
```suggestion
        await bot.add_cog(Suggestions(bot, key["webhook_url"]))
```

File: types_/config.py [hunk 11]
```diff
@@ -31,6 +31,10 @@ class BadBin(TypedDict):
     domains: list[str]
 
 
+class Suggestions(TypedDict):
+    url: str
```
Comment (AbstractUmbra) [inline types_/config.py:35]:
```suggestion
    webhook_url: str
```

**Review by AbstractUmbra — CHANGES_REQUESTED** [review]

File: launcher.py
Comment (Duued) [reply launcher.py:1]: An accident. I think it's back

File: config.template.toml [hunk 12]
```diff
@@ -24,6 +24,9 @@ webhook_avatar_url = "" # optional
 runner = 123456789 # optional: sets the webhook avatar url (will be overridden by the webhook_avatar_url attribute)
 level = 20
 
+[SUGGESTIONS] # optional
+webhook_url = "" # optional
```
Comment (AbstractUmbra) [inline config.template.toml:28]:
```suggestion
webhook_url = ""
```

**Review by AbstractUmbra — APPROVED** [review]
Last item but I'll get it.
